src/Core/IdentityServer/CustomTokenRequestValidator.cs

```diff
@@ -100,6 +83,37 @@ protected async override Task<(User, bool)> ValidateContextAsync(CustomTokenRequ
                     context.Result.ValidatedRequest.ClientClaims.Add(claim);
                 }
             }
+
+            if (context.Result.CustomResponse == null || user.MasterPassword != null)
+            {
+                return;
+            }
+
+            // KeyConnector responses below
+
+            if (context.Result.ValidatedRequest.GrantType == "client_credentials" &&
+                user.UsesKeyConnector)
+            {
+                // Apikey login
+                // KeyConnectorUrl is configured in the CLI client, just disable master password reset    
+                context.Result.CustomResponse["ResetMasterPassword"] = false;
+                return;
+            }
+
+            // SSO login
+            var organizationClaim = context.Result.ValidatedRequest.Subject?.FindFirst(c => c.Type == "organizationId");
```

Comment:
Do we need to add null checks in this section, e.g. what if the user is not using SSO and needs to reset their password? I assume they'll pass the guard on line 87 and then encounter errors below.

Reply:
Fair point, added null check.